# Exploring eBay Car Sales Data

We want to clean and analyze data of used car listings from a dataset from the classifieds section of the German eBay website 

Kleinanzeigen.

In [2]:
import os

os.getcwd()

'C:\\Users\\caspe'

In [3]:
os.chdir('C:\\Users\\caspe\\Downloads')

In [4]:
import pandas as pd
import numpy as np

autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [5]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [6]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [7]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

- The dataset has 20 columns and 50,000 rows
- The datatype of the dataset is mainly object (text/string)
- The column names are majorly in camel case rather than the preferred snake case so spaces cannot easily be replaced by underscore
- Some columns have null values

In [8]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

Let's change the column names from camel case to snake case and also convert some of the names to more descriptive names

In [9]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest', 'vehicle_type',
                   'registration_year', 'gearbox', 'power_ps', 'model', 'odometer', 'registration_month',
                    'fuel_type', 'brand', 'unrepaired_damage', 'ad_created', 'no_of_pictures', 'postal_code', 'last_seen']

In [10]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  no_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [11]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  no_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

- The seller and offer_type columns can be discarded because they contain almost similar data throughout
- The no_of_pictures column can be dropped as their are no pictures for any of the rows
- The min for registration_month column is 0, which is impossible and indicates inaccurate data
- The max for registration_year is higher than the value for 75%, this is abnormal and is indicative of inaccurate data
- Some columns have data stored as text rather than numeric data, e.g price and odometer

In [12]:
autos['price'].value_counts()

$0         1421
$500        781
$1,500      734
$2,500      643
$1,000      639
           ... 
$414          1
$79,933       1
$5,198        1
$18,890       1
$16,995       1
Name: price, Length: 2357, dtype: int64

In [13]:
autos['price'].dropna(inplace=True)

In [14]:
autos['price'].head()

0    $5,000
1    $8,500
2    $8,990
3    $4,350
4    $1,350
Name: price, dtype: object

In [15]:
autos['price'] = autos['price'].str.replace('$','').str.replace(',','').astype(int)

C:\Users\caspe\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  """Entry point for launching an IPython kernel.


In [16]:
autos['price'].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int32

In [17]:
autos['odometer'].dropna(inplace=True)

In [18]:
autos['odometer'].head()

0    150,000km
1    150,000km
2     70,000km
3     70,000km
4    150,000km
Name: odometer, dtype: object

In [19]:
autos['odometer'] = autos['odometer'].str.replace(',','').str.replace('km','').astype(int)

In [20]:
autos['odometer'].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer, dtype: int32

In [21]:
autos = autos.rename({'odometer':'odometer_km'}, axis=1)

In [22]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere       150000                   3       lpg   
1  automatik       286     7er       150000                   6    benzin   
2    manuell       102    golf        70000                   7    benzin   
3  automatik        71  fortwo        70000                   6    benzin   
4    manuell         0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  no_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [23]:
autos['price'].unique().shape

(2357,)

In [24]:
autos['odometer_km'].unique().shape

(13,)

In [25]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [26]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [27]:
autos['odometer_km'].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

As seen above the values in the odometer_km column seem to show consistency in their differences and column seems not to have any high or low outliers.

In [28]:
autos['price'].value_counts().sort_index(ascending=True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64

In [29]:
autos['price'].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

From the above it can be seen that the price has high outliers after 350,000, so it is safe to say the values after 350,000 can be removed  from the dataset.

In [30]:
autos['odometer_km'].value_counts().sort_index(ascending=False).head(10)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
Name: odometer_km, dtype: int64

We can now remove outliers from the price column

In [31]:
autos[autos["price"].between(1,350000)]

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offer_type  price   abtest vehicle_type  registration_year  \
0      privat    Angebot   5000  control          bus               2004   
1      privat    Angebot   8500  control    limousine               1997   
2      privat    Angebot   8990     test    limousine               2009   
3      privat    Angebot   4350  control   kleinwagen               2007   
4      privat    Angebot   1350     test        kombi               2003   
...       ...        ...    ...      ...          ...                ...   
49995  privat    Angebot  24900  control    limousine               2011   
49996  privat    Angebot   1980  control       cabrio               1996   
49997  privat    Angebot  13200     test       cabrio               2014   
49998  privat    Angebot  22900  control        kombi               2013   
49999  privat    Angebot   1250  control    limousine               1996   

         gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0        manuell       158  andere       150000                   3       lpg   
1      automatik       286     7er       150000                   6    benzin   
2        manuell       102    golf        70000                   7    benzin   
3      automatik        71  fortwo        70000                   6    benzin   
4        manuell         0   focus       150000                   7    benzin   
...          ...       ...     ...          ...                 ...       ...   
49995  automatik       239      q5       100000                   1    diesel   
49996    manuell        75   astra       150000                   5    benzin   
49997  automatik        69     500         5000                  11    benzin   
49998    manuell       150      a3        40000                  11    diesel   
49999    manuell       101  vectra       150000                   1    benzin   

            brand unrepaired_damage           ad_created  no_of_pictures  \
0         peugeot              nein  2016-03-26 00:00:00               0   
1             bmw              nein  2016-04-04 00:00:00               0   
2      volkswagen              nein  2016-03-26 00:00:00               0   
3           smart              nein  2016-03-12 00:00:00               0   
4            ford              nein  2016-04-01 00:00:00               0   
...           ...               ...                  ...             ...   
49995        audi              nein  2016-03-27 00:00:00               0   
49996        opel              nein  2016-03-28 00:00:00               0   
49997        fiat              nein  2016-04-02 00:00:00               0   
49998        audi              nein  2016-03-08 00:00:00               0   
49999        opel              nein  2016-03-13 00:00:00               0   

       postal_code            last_seen  
0            79588  2016-04-06 06:45:54  
1            71034  2016-04-06 14:45:08  
2            35394  2016-04-06 20:15:37  
3            33729  2016-03-15 03:16:28  
4            39218  2016-04-01 1

In [32]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere       150000                   3       lpg   
1  automatik       286     7er       150000                   6    benzin   
2    manuell       102    golf        70000                   7    benzin   
3  automatik        71  fortwo        70000                   6    benzin   
4    manuell         0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  no_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [33]:
autos['price'].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int32

Converting the date_crawled, last_seen and ad_created columns to numerical representation

In [34]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [35]:
autos['date_crawled'].value_counts()

2016-04-02 11:37:04    3
2016-03-30 19:48:02    3
2016-04-02 15:49:30    3
2016-03-22 09:51:06    3
2016-03-05 16:57:05    3
                      ..
2016-03-18 22:07:04    1
2016-03-28 07:36:23    1
2016-03-31 22:49:18    1
2016-04-02 19:50:18    1
2016-03-14 00:42:12    1
Name: date_crawled, Length: 48213, dtype: int64

In [36]:
print(autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index())

2016-03-05    0.02538
2016-03-06    0.01394
2016-03-07    0.03596
2016-03-08    0.03330
2016-03-09    0.03322
2016-03-10    0.03212
2016-03-11    0.03248
2016-03-12    0.03678
2016-03-13    0.01556
2016-03-14    0.03662
2016-03-15    0.03398
2016-03-16    0.02950
2016-03-17    0.03152
2016-03-18    0.01306
2016-03-19    0.03490
2016-03-20    0.03782
2016-03-21    0.03752
2016-03-22    0.03294
2016-03-23    0.03238
2016-03-24    0.02910
2016-03-25    0.03174
2016-03-26    0.03248
2016-03-27    0.03104
2016-03-28    0.03484
2016-03-29    0.03418
2016-03-30    0.03362
2016-03-31    0.03192
2016-04-01    0.03380
2016-04-02    0.03540
2016-04-03    0.03868
2016-04-04    0.03652
2016-04-05    0.01310
2016-04-06    0.00318
2016-04-07    0.00142
Name: date_crawled, dtype: float64


In [37]:
print(autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index())

2015-06-11    0.00002
2015-08-10    0.00002
2015-09-09    0.00002
2015-11-10    0.00002
2015-12-05    0.00002
               ...   
2016-04-03    0.03892
2016-04-04    0.03688
2016-04-05    0.01184
2016-04-06    0.00326
2016-04-07    0.00128
Name: ad_created, Length: 76, dtype: float64


In [38]:
print(autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index())

2016-03-05    0.00108
2016-03-06    0.00442
2016-03-07    0.00536
2016-03-08    0.00760
2016-03-09    0.00986
2016-03-10    0.01076
2016-03-11    0.01252
2016-03-12    0.02382
2016-03-13    0.00898
2016-03-14    0.01280
2016-03-15    0.01588
2016-03-16    0.01644
2016-03-17    0.02792
2016-03-18    0.00742
2016-03-19    0.01574
2016-03-20    0.02070
2016-03-21    0.02074
2016-03-22    0.02158
2016-03-23    0.01858
2016-03-24    0.01956
2016-03-25    0.01920
2016-03-26    0.01696
2016-03-27    0.01602
2016-03-28    0.02086
2016-03-29    0.02234
2016-03-30    0.02484
2016-03-31    0.02384
2016-04-01    0.02310
2016-04-02    0.02490
2016-04-03    0.02536
2016-04-04    0.02462
2016-04-05    0.12428
2016-04-06    0.22100
2016-04-07    0.13092
Name: last_seen, dtype: float64


In [39]:
autos['registration_year'].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The registration_year column seems to have some odd values as shown by the min and max the min value is 1000 way before the invention of cars and the max value is 9999 way into the future. 

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.

In [40]:
autos[autos["registration_year"].between(1900,2016)]

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offer_type  price   abtest vehicle_type  registration_year  \
0      privat    Angebot   5000  control          bus               2004   
1      privat    Angebot   8500  control    limousine               1997   
2      privat    Angebot   8990     test    limousine               2009   
3      privat    Angebot   4350  control   kleinwagen               2007   
4      privat    Angebot   1350     test        kombi               2003   
...       ...        ...    ...      ...          ...                ...   
49995  privat    Angebot  24900  control    limousine               2011   
49996  privat    Angebot   1980  control       cabrio               1996   
49997  privat    Angebot  13200     test       cabrio               2014   
49998  privat    Angebot  22900  control        kombi               2013   
49999  privat    Angebot   1250  control    limousine               1996   

         gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0        manuell       158  andere       150000                   3       lpg   
1      automatik       286     7er       150000                   6    benzin   
2        manuell       102    golf        70000                   7    benzin   
3      automatik        71  fortwo        70000                   6    benzin   
4        manuell         0   focus       150000                   7    benzin   
...          ...       ...     ...          ...                 ...       ...   
49995  automatik       239      q5       100000                   1    diesel   
49996    manuell        75   astra       150000                   5    benzin   
49997  automatik        69     500         5000                  11    benzin   
49998    manuell       150      a3        40000                  11    diesel   
49999    manuell       101  vectra       150000                   1    benzin   

            brand unrepaired_damage           ad_created  no_of_pictures  \
0         peugeot              nein  2016-03-26 00:00:00               0   
1             bmw              nein  2016-04-04 00:00:00               0   
2      volkswagen              nein  2016-03-26 00:00:00               0   
3           smart              nein  2016-03-12 00:00:00               0   
4            ford              nein  2016-04-01 00:00:00               0   
...           ...               ...                  ...             ...   
49995        audi              nein  2016-03-27 00:00:00               0   
49996        opel              nein  2016-03-28 00:00:00               0   
49997        fiat              nein  2016-04-02 00:00:00               0   
49998        audi              nein  2016-03-08 00:00:00               0   
49999        opel              nein  2016-03-13 00:00:00               0   

       postal_code            last_seen  
0            79588  2016-04-06 06:45:54  
1            71034  2016-04-06 14:45:08  
2            35394  2016-04-06 20:15:37  
3            33729  2016-03-15 03:16:28  
4            39218  2016-04-01 1

In [41]:
autos['registration_year'].value_counts(normalize=True).head()

2000    0.06708
2005    0.06030
1999    0.06000
2004    0.05474
2003    0.05454
Name: registration_year, dtype: float64

We want to perform some aggregation operations on the brand column 

In [42]:
brands = autos['brand'].unique()

In [43]:
brands

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [44]:
brands = autos['brand'].value_counts(normalize=True)

In [45]:
brands

volkswagen        0.21374
opel              0.10922
bmw               0.10858
mercedes_benz     0.09468
audi              0.08566
ford              0.06958
renault           0.04808
peugeot           0.02912
fiat              0.02616
seat              0.01882
skoda             0.01572
mazda             0.01514
nissan            0.01508
smart             0.01402
citroen           0.01402
toyota            0.01234
sonstige_autos    0.01092
hyundai           0.00976
volvo             0.00914
mini              0.00848
mitsubishi        0.00812
honda             0.00798
kia               0.00712
alfa_romeo        0.00658
porsche           0.00588
suzuki            0.00586
chevrolet         0.00566
chrysler          0.00362
dacia             0.00258
daihatsu          0.00256
jeep              0.00220
subaru            0.00218
land_rover        0.00198
saab              0.00160
daewoo            0.00158
trabant           0.00156
jaguar            0.00154
rover             0.00138
lancia      

German cars seem to be the more popular cars (top 5) with Volkswagen being the most popular, we'll limit our analysis to brands with more than 5% of the total sales data.

In [46]:
common_brands = brands[brands > 0.05].index

In [47]:
common_brands

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford'], dtype='object')

In [48]:
# Create an empty dictionary
popular_brands = {}

for brands in common_brands:
    selected_brands = autos[autos['brand'] == brands]
    mean = selected_brands['price'].mean()
    popular_brands[brands] = int(mean)

In [49]:
popular_brands

{'volkswagen': 6384,
 'opel': 5106,
 'bmw': 8252,
 'mercedes_benz': 29511,
 'audi': 8965,
 'ford': 7105}

Based on the top 5 brands, Mercedes Benz seems to be the most expensive brand followed by Audi, then BMW, Opel is the least expensive and Volkswagen seems to be in-between hence could be the reason for it's popularity.

In [50]:
mean_mileage = {}

for brands in common_brands:
    selected_brands = autos[autos['brand'] == brands]
    mileage = selected_brands['odometer_km'].mean()
    mean_mileage[brands] = int(mileage)

In [51]:
mean_mileage

{'volkswagen': 128955,
 'opel': 129298,
 'bmw': 132521,
 'mercedes_benz': 130886,
 'audi': 129643,
 'ford': 124131}

In [52]:
pb_series = pd.Series(popular_brands).sort_values(ascending=False)
print(pb_series)

mercedes_benz    29511
audi              8965
bmw               8252
ford              7105
volkswagen        6384
opel              5106
dtype: int64


In [53]:
df_mean = pd.DataFrame(pb_series, columns=['mean'])
df_mean

mean
mercedes_benz  29511
audi            8965
bmw             8252
ford            7105
volkswagen      6384
opel            5106

In [54]:
mm_series = pd.Series(mean_mileage).sort_values(ascending=False)
print(mm_series)

bmw              132521
mercedes_benz    130886
audi             129643
opel             129298
volkswagen       128955
ford             124131
dtype: int64


In [55]:
df_mileage = pd.DataFrame(mm_series, columns=['mileage'])
df_mileage

mileage
bmw             132521
mercedes_benz   130886
audi            129643
opel            129298
volkswagen      128955
ford            124131

In [56]:
df_mileage['mean'] = df_mean
df_mileage

mileage   mean
bmw             132521   8252
mercedes_benz   130886  29511
audi            129643   8965
opel            129298   5106
volkswagen      128955   6384
ford            124131   7105

Comparing both DataFrames it can be observed that the mileage seems to be related to the price of each brand, as the brands with the highest mileage are also the pricier brands i.e. Mercedes Benz, BMW and Audi. 

With Opel seeming to be the only outlier. 

In [57]:
autos.head(20)

date_crawled                                               name  \
0   2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1   2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2   2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3   2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4   2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5   2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6   2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7   2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8   2016-03-22 16:51:34                                         Seat_Arosa   
9   2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10  2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11  2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12  2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13  2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14  2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15  2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16  2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17  2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18  2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19  2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   

    seller offer_type  price   abtest vehicle_type  registration_year  \
0   privat    Angebot   5000  control          bus               2004   
1   privat    Angebot   8500  control    limousine               1997   
2   privat    Angebot   8990     test    limousine               2009   
3   privat    Angebot   4350  control   kleinwagen               2007   
4   privat    Angebot   1350     test        kombi               2003   
5   privat    Angebot   7900     test          bus               2006   
6   privat    Angebot    300     test    limousine               1995   
7   privat    Angebot   1990  control    limousine               1998   
8   privat    Angebot    250     test          NaN               2000   
9   privat    Angebot    590  control          bus               1997   
10  privat    Angebot    999     test          NaN               2017   
11  privat    Angebot    350  control          NaN               2000   
12  privat    Angebot   5299  control   kleinwagen               2010   
13  privat    Angebot   1350  control    limousine               1999   
14  privat    Angebot   3999     test   kleinwagen               2007   
15  privat    Angebot  18900     test        coupe               1982   
16  privat    Angebot    350     test        kombi               1999   
17  privat    Angebot   5500     test        coupe               1990   
18  privat    Angebot    300  control          bus               1995   
19  privat    Angebot   4150  control          suv               2004   

      gearbox  power_ps     model  odometer_km  registration_month fuel_type  \
0     manuell       158    andere       150000                   3       lpg   
1   automatik       286       7er       150000                   6    benzin   
2     manuell       102      golf        70000                   7    benzin   
3   automatik        71    fortwo        70000                   6    benzin   
4     manuell         0     focus       150000                   7    benzin   
5   automatik       150   voyager       150000                   4    diesel   
6     manuell        90      golf       150000                   8    benzin   
7     manuell        90      golf       150000                  12    diesel   
8     manuell         0     arosa       150000                  10       NaN   
9     manuell        90    megane     

In [58]:
autos.tail()

date_crawled                                               name  \
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offer_type  price   abtest vehicle_type  registration_year  \
49995  privat    Angebot  24900  control    limousine               2011   
49996  privat    Angebot   1980  control       cabrio               1996   
49997  privat    Angebot  13200     test       cabrio               2014   
49998  privat    Angebot  22900  control        kombi               2013   
49999  privat    Angebot   1250  control    limousine               1996   

         gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
49995  automatik       239      q5       100000                   1    diesel   
49996    manuell        75   astra       150000                   5    benzin   
49997  automatik        69     500         5000                  11    benzin   
49998    manuell       150      a3        40000                  11    diesel   
49999    manuell       101  vectra       150000                   1    benzin   

      brand unrepaired_damage           ad_created  no_of_pictures  \
49995  audi              nein  2016-03-27 00:00:00               0   
49996  opel              nein  2016-03-28 00:00:00               0   
49997  fiat              nein  2016-04-02 00:00:00               0   
49998  audi              nein  2016-03-08 00:00:00               0   
49999  opel              nein  2016-03-13 00:00:00               0   

       postal_code            last_seen  
49995        82131  2016-04-01 13:47:40  
49996        44807  2016-04-02 14:18:02  
49997        73430  2016-04-04 11:47:27  
49998        35683  2016-04-05 16:45:07  
49999        45897  2016-04-06 21:18:48

In [59]:
autos['seller'].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [60]:
autos['gearbox'].value_counts()

manuell      36993
automatik    10327
Name: gearbox, dtype: int64

In [61]:
autos['fuel_type'].value_counts()

benzin     30107
diesel     14567
lpg          691
cng           75
hybrid        37
andere        22
elektro       19
Name: fuel_type, dtype: int64

In [62]:
autos['unrepaired_damage'].value_counts()

nein    35232
ja       4939
Name: unrepaired_damage, dtype: int64

In [63]:
autos['offer_type'].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

In [64]:
autos['vehicle_type'].value_counts()

limousine     12859
kleinwagen    10822
kombi          9127
bus            4093
cabrio         3061
coupe          2537
suv            1986
andere          420
Name: vehicle_type, dtype: int64

In [65]:
autos['seller'] = autos['seller'].str.replace('privat', 'private').str.replace('gewerblich', 'commercial')

In [66]:
autos['seller'].head()

0    private
1    private
2    private
3    private
4    private
Name: seller, dtype: object

In [67]:
autos['gearbox'] = autos['gearbox'].str.replace('manuell', 'manual').str.replace('automatik', 'automatic')

In [68]:
autos['gearbox'].head()

0       manual
1    automatic
2       manual
3    automatic
4       manual
Name: gearbox, dtype: object

In [69]:
autos['unrepaired_damage'] = autos['unrepaired_damage'].str.replace('nein', 'no').str.replace('ja', 'yes')

In [70]:
autos['unrepaired_damage'].head()

0    no
1    no
2    no
3    no
4    no
Name: unrepaired_damage, dtype: object

In [71]:
autos['vehicle_type'] = autos['vehicle_type'].str.replace('kleinwagen', 'small car').str.replace('kombi', 'station wagon').str.replace('cabrio', 'convertible').str.replace('andere', 'others')

In [72]:
autos['vehicle_type'].head()

0              bus
1        limousine
2        limousine
3        small car
4    station wagon
Name: vehicle_type, dtype: object

In [73]:
autos['fuel_type'] = autos['fuel_type'].str.replace('andere', 'others').str.replace('elektro', 'electro')

In [74]:
autos['fuel_type'].head()

0       lpg
1    benzin
2    benzin
3    benzin
4    benzin
Name: fuel_type, dtype: object

In [75]:
autos.head(20)

date_crawled                                               name  \
0   2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1   2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2   2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3   2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4   2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5   2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6   2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7   2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8   2016-03-22 16:51:34                                         Seat_Arosa   
9   2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10  2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11  2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12  2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13  2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14  2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15  2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16  2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17  2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18  2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19  2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   

     seller offer_type  price   abtest   vehicle_type  registration_year  \
0   private    Angebot   5000  control            bus               2004   
1   private    Angebot   8500  control      limousine               1997   
2   private    Angebot   8990     test      limousine               2009   
3   private    Angebot   4350  control      small car               2007   
4   private    Angebot   1350     test  station wagon               2003   
5   private    Angebot   7900     test            bus               2006   
6   private    Angebot    300     test      limousine               1995   
7   private    Angebot   1990  control      limousine               1998   
8   private    Angebot    250     test            NaN               2000   
9   private    Angebot    590  control            bus               1997   
10  private    Angebot    999     test            NaN               2017   
11  private    Angebot    350  control            NaN               2000   
12  private    Angebot   5299  control      small car               2010   
13  private    Angebot   1350  control      limousine               1999   
14  private    Angebot   3999     test      small car               2007   
15  private    Angebot  18900     test          coupe               1982   
16  private    Angebot    350     test  station wagon               1999   
17  private    Angebot   5500     test          coupe               1990   
18  private    Angebot    300  control            bus               1995   
19  private    Angebot   4150  control            suv               2004   

      gearbox  power_ps     model  odometer_km  registration_month fuel_type  \
0      manual       158    andere       150000                   3       lpg   
1   automatic       286       7er       150000                   6    benzin   
2      manual       102      golf        70000                   7    benzin   
3   automatic        71    fortwo        70000                   6    benzin   
4      manual         0     focus       150000                   7    benzin   
5   automatic       150   voyager       150000                   4    diesel   
6      manual        90      golf       150000                   8    benzin   
7      manual        90      golf       150000                  12    diesel   
8      manual         0     arosa       150000         

In [76]:
autos['date_crawled'].dropna(inplace=True)

In [77]:
autos['date_crawled'].head()

0    2016-03-26 17:47:46
1    2016-04-04 13:38:56
2    2016-03-26 18:57:24
3    2016-03-12 16:58:10
4    2016-04-01 14:38:50
Name: date_crawled, dtype: object

In [79]:
autos_date = autos['date_crawled'].str.split(' ')

In [80]:
autos_date.head()

0    [2016-03-26, 17:47:46]
1    [2016-04-04, 13:38:56]
2    [2016-03-26, 18:57:24]
3    [2016-03-12, 16:58:10]
4    [2016-04-01, 14:38:50]
Name: date_crawled, dtype: object

In [89]:
date_crawled = []

for item in autos_date:
    date_crawled.append(item[0])
print(date_crawled[:5])

['2016-03-26', '2016-04-04', '2016-03-26', '2016-03-12', '2016-04-01']


In [97]:
autos['date_crawled'] = date_crawled
autos['date_crawled'].head(10)

0    2016-03-26
1    2016-04-04
2    2016-03-26
3    2016-03-12
4    2016-04-01
5    2016-03-21
6    2016-03-20
7    2016-03-16
8    2016-03-22
9    2016-03-16
Name: date_crawled, dtype: object

In [98]:
autos.head()

date_crawled                                               name   seller  \
0   2016-03-26                   Peugeot_807_160_NAVTECH_ON_BOARD  private   
1   2016-04-04         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  private   
2   2016-03-26                         Volkswagen_Golf_1.6_United  private   
3   2016-03-12  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...  private   
4   2016-04-01  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...  private   

  offer_type  price   abtest   vehicle_type  registration_year    gearbox  \
0    Angebot   5000  control            bus               2004     manual   
1    Angebot   8500  control      limousine               1997  automatic   
2    Angebot   8990     test      limousine               2009     manual   
3    Angebot   4350  control      small car               2007  automatic   
4    Angebot   1350     test  station wagon               2003     manual   

   power_ps   model  odometer_km  registration_month fuel_type       brand  \
0       158  andere       150000                   3       lpg     peugeot   
1       286     7er       150000                   6    benzin         bmw   
2       102    golf        70000                   7    benzin  volkswagen   
3        71  fortwo        70000                   6    benzin       smart   
4         0   focus       150000                   7    benzin        ford   

  unrepaired_damage           ad_created  no_of_pictures  postal_code  \
0                no  2016-03-26 00:00:00               0        79588   
1                no  2016-04-04 00:00:00               0        71034   
2                no  2016-03-26 00:00:00               0        35394   
3                no  2016-03-12 00:00:00               0        33729   
4                no  2016-04-01 00:00:00               0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

In [101]:
autos['date_crawled'] = autos['date_crawled'].str.replace('-','').astype(int)
autos['date_crawled'].head()

0    20160326
1    20160404
2    20160326
3    20160312
4    20160401
Name: date_crawled, dtype: int32

In [102]:
autos.head()

date_crawled                                               name   seller  \
0      20160326                   Peugeot_807_160_NAVTECH_ON_BOARD  private   
1      20160404         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  private   
2      20160326                         Volkswagen_Golf_1.6_United  private   
3      20160312  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...  private   
4      20160401  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...  private   

  offer_type  price   abtest   vehicle_type  registration_year    gearbox  \
0    Angebot   5000  control            bus               2004     manual   
1    Angebot   8500  control      limousine               1997  automatic   
2    Angebot   8990     test      limousine               2009     manual   
3    Angebot   4350  control      small car               2007  automatic   
4    Angebot   1350     test  station wagon               2003     manual   

   power_ps   model  odometer_km  registration_month fuel_type       brand  \
0       158  andere       150000                   3       lpg     peugeot   
1       286     7er       150000                   6    benzin         bmw   
2       102    golf        70000                   7    benzin  volkswagen   
3        71  fortwo        70000                   6    benzin       smart   
4         0   focus       150000                   7    benzin        ford   

  unrepaired_damage           ad_created  no_of_pictures  postal_code  \
0                no  2016-03-26 00:00:00               0        79588   
1                no  2016-04-04 00:00:00               0        71034   
2                no  2016-03-26 00:00:00               0        35394   
3                no  2016-03-12 00:00:00               0        33729   
4                no  2016-04-01 00:00:00               0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

In [103]:
autos['ad_created'].dropna(inplace=True)

In [104]:
autos['ad_created'].head()

0    2016-03-26 00:00:00
1    2016-04-04 00:00:00
2    2016-03-26 00:00:00
3    2016-03-12 00:00:00
4    2016-04-01 00:00:00
Name: ad_created, dtype: object

In [105]:
ad_date = autos['ad_created'].str.split(' ')
ad_date

0        [2016-03-26, 00:00:00]
1        [2016-04-04, 00:00:00]
2        [2016-03-26, 00:00:00]
3        [2016-03-12, 00:00:00]
4        [2016-04-01, 00:00:00]
                  ...          
49995    [2016-03-27, 00:00:00]
49996    [2016-03-28, 00:00:00]
49997    [2016-04-02, 00:00:00]
49998    [2016-03-08, 00:00:00]
49999    [2016-03-13, 00:00:00]
Name: ad_created, Length: 50000, dtype: object

In [107]:
ad_created = []

for item in ad_date:
    ad_created.append(item[0])
print(ad_created[:5])

['2016-03-26', '2016-04-04', '2016-03-26', '2016-03-12', '2016-04-01']


In [108]:
autos['ad_created'] = ad_created
autos['ad_created'].head()

0    2016-03-26
1    2016-04-04
2    2016-03-26
3    2016-03-12
4    2016-04-01
Name: ad_created, dtype: object

In [110]:
autos['ad_created'] = autos['ad_created'].str.replace('-','').astype(int)
autos['ad_created'].head()

0    20160326
1    20160404
2    20160326
3    20160312
4    20160401
Name: ad_created, dtype: int32

In [111]:
autos.head()

date_crawled                                               name   seller  \
0      20160326                   Peugeot_807_160_NAVTECH_ON_BOARD  private   
1      20160404         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  private   
2      20160326                         Volkswagen_Golf_1.6_United  private   
3      20160312  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...  private   
4      20160401  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...  private   

  offer_type  price   abtest   vehicle_type  registration_year    gearbox  \
0    Angebot   5000  control            bus               2004     manual   
1    Angebot   8500  control      limousine               1997  automatic   
2    Angebot   8990     test      limousine               2009     manual   
3    Angebot   4350  control      small car               2007  automatic   
4    Angebot   1350     test  station wagon               2003     manual   

   power_ps   model  odometer_km  registration_month fuel_type       brand  \
0       158  andere       150000                   3       lpg     peugeot   
1       286     7er       150000                   6    benzin         bmw   
2       102    golf        70000                   7    benzin  volkswagen   
3        71  fortwo        70000                   6    benzin       smart   
4         0   focus       150000                   7    benzin        ford   

  unrepaired_damage  ad_created  no_of_pictures  postal_code  \
0                no    20160326               0        79588   
1                no    20160404               0        71034   
2                no    20160326               0        35394   
3                no    20160312               0        33729   
4                no    20160401               0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

In [112]:
autos['last_seen'].dropna(inplace=True)

In [113]:
last_date = autos['last_seen'].str.split(' ')
last_date

0        [2016-04-06, 06:45:54]
1        [2016-04-06, 14:45:08]
2        [2016-04-06, 20:15:37]
3        [2016-03-15, 03:16:28]
4        [2016-04-01, 14:38:50]
                  ...          
49995    [2016-04-01, 13:47:40]
49996    [2016-04-02, 14:18:02]
49997    [2016-04-04, 11:47:27]
49998    [2016-04-05, 16:45:07]
49999    [2016-04-06, 21:18:48]
Name: last_seen, Length: 50000, dtype: object

In [114]:
last_seen = []

for item in last_date:
    last_seen.append(item[0])
print(last_seen[:5])

['2016-04-06', '2016-04-06', '2016-04-06', '2016-03-15', '2016-04-01']


In [115]:
autos['last_seen'] = last_seen
autos['last_seen'].head()

0    2016-04-06
1    2016-04-06
2    2016-04-06
3    2016-03-15
4    2016-04-01
Name: last_seen, dtype: object

In [116]:
autos['last_seen'] = autos['last_seen'].str.replace('-','').astype(int)
autos['last_seen'].head()

0    20160406
1    20160406
2    20160406
3    20160315
4    20160401
Name: last_seen, dtype: int32

In [117]:
autos.head()

date_crawled                                               name   seller  \
0      20160326                   Peugeot_807_160_NAVTECH_ON_BOARD  private   
1      20160404         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  private   
2      20160326                         Volkswagen_Golf_1.6_United  private   
3      20160312  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...  private   
4      20160401  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...  private   

  offer_type  price   abtest   vehicle_type  registration_year    gearbox  \
0    Angebot   5000  control            bus               2004     manual   
1    Angebot   8500  control      limousine               1997  automatic   
2    Angebot   8990     test      limousine               2009     manual   
3    Angebot   4350  control      small car               2007  automatic   
4    Angebot   1350     test  station wagon               2003     manual   

   power_ps   model  odometer_km  registration_month fuel_type       brand  \
0       158  andere       150000                   3       lpg     peugeot   
1       286     7er       150000                   6    benzin         bmw   
2       102    golf        70000                   7    benzin  volkswagen   
3        71  fortwo        70000                   6    benzin       smart   
4         0   focus       150000                   7    benzin        ford   

  unrepaired_damage  ad_created  no_of_pictures  postal_code  last_seen  
0                no    20160326               0        79588   20160406  
1                no    20160404               0        71034   20160406  
2                no    20160326               0        35394   20160406  
3                no    20160312               0        33729   20160315  
4                no    20160401               0        39218   20160401